File: src/me/anno/utils/pooling/JomlPools.java
package me.anno.utils.pooling;

import org.joml.*;

// can be converted to Kotlin without any issues
// is in Java to prevent a lot of tiny classes from being created for lambdas, and objects (that's the worst part...)
// (small optimization for JVM2WASM, might help JVM as well)
public class JomlPools {

    public static final Stack<Vector2i> vec2i = new Stack<>(Vector2i.class);
    public static final Stack<Vector2f> vec2f = new Stack<>(Vector2f.class);
    public static final Stack<Vector2d> vec2d = new Stack<>(Vector2d.class);

    public static final Stack<Vector3i> vec3i = new Stack<>(Vector3i.class);
    public static final Stack<Vector3f> vec3f = new Stack<>(Vector3f.class);
    public static final Stack<Vector3d> vec3d = new Stack<>(Vector3d.class);

    public static final Stack<Vector4i> vec4i = new Stack<>(Vector4i.class);
    public static final Stack<Vector4f> vec4f = new Stack<>(Vector4f.class);
    public static final Stack<Vector4d> vec4d = new Stack<>(Vector4d.class);

    public static final Stack<Quaternionf> quat4f = new Stack<>(Quaternionf.class);
    public static final Stack<Quaterniond> quat4d = new Stack<>(Quaterniond.class);

    public static final Stack<Matrix2f> mat2f = new Stack<>(Matrix2f.class);
    public static final Stack<Matrix2d> mat2d = new Stack<>(Matrix2d.class);

    public static final Stack<Matrix3f> mat3f = new Stack<>(Matrix3f.class);
    public static final Stack<Matrix3d> mat3d = new Stack<>(Matrix3d.class);

    public static final Stack<Matrix4f> mat4f = new Stack<>(Matrix4f.class);
    public static final Stack<Matrix4d> mat4d = new Stack<>(Matrix4d.class);

    public static final Stack<Matrix4x3f> mat4x3f = new Stack<>(Matrix4x3f.class);
    public static final Stack<Matrix4x3d> mat4x3d = new Stack<>(Matrix4x3d.class);

    public static final Stack<AABBf> aabbf = new Stack<>(AABBf.class);
    public static final Stack<AABBd> aabbd = new Stack<>(AABBd.class);

    public static void reset() {

        vec2i.reset();
        vec2f.reset();
        vec2d.reset();

        vec3i.reset();
        vec3f.reset();
        vec3d.reset();

        vec4i.reset();
        vec4f.reset();
        vec4d.reset();

        quat4f.reset();
        quat4d.reset();

        mat2f.reset();
        mat2d.reset();

        mat3f.reset();
        mat3d.reset();

        mat4f.reset();
        mat4d.reset();

        mat4x3f.reset();
        mat4x3d.reset();

        aabbf.reset();
        aabbd.reset();

    }

}

File: src/me/anno/utils/pooling/JomlPools.kt
package me.anno.utils.pooling

import org.joml.AABBd
import org.joml.AABBf
import org.joml.Matrix2d
import org.joml.Matrix2f
import org.joml.Matrix3d
import org.joml.Matrix3f
import org.joml.Matrix4d
import org.joml.Matrix4f
import org.joml.Matrix4x3d
import org.joml.Matrix4x3f
import org.joml.Quaterniond
import org.joml.Quaternionf
import org.joml.Vector2d
import org.joml.Vector2f
import org.joml.Vector2i
import org.joml.Vector3d
import org.joml.Vector3f
import org.joml.Vector3i
import org.joml.Vector4d
import org.joml.Vector4f
import org.joml.Vector4i

object JomlPools {
    val vec2i = Stack(Vector2i::class)
    val vec2f = Stack(Vector2f::class)
    val vec2d = Stack(Vector2d::class)
    val vec3i = Stack(Vector3i::class)
    val vec3f = Stack(Vector3f::class)
    val vec3d = Stack(Vector3d::class)
    val vec4i = Stack(Vector4i::class)
    val vec4f = Stack(Vector4f::class)
    val vec4d = Stack(Vector4d::class)
    val quat4f = Stack(Quaternionf::class)
    val quat4d = Stack(Quaterniond::class)
    val mat2f = Stack(Matrix2f::class)
    val mat2d = Stack(Matrix2d::class)
    val mat3f = Stack(Matrix3f::class)
    val mat3d = Stack(Matrix3d::class)
    val mat4f = Stack(Matrix4f::class)
    val mat4d = Stack(Matrix4d::class)
    val mat4x3f = Stack(Matrix4x3f::class)
    val mat4x3d = Stack(Matrix4x3d::class)
    val aabbf = Stack(AABBf::class)
    val aabbd = Stack(AABBd::class)
}